File: App/Resources/AlgorithmDetails/binaryquicksortrecursive/binaryquicksortrecursive.java
import java.util.Arrays;

public class binaryquicksortrecursive {
  private static int mostSignificantBit(int value) {
    if (value == 0) {
      return -1;
    }
    int bit = 0;
    while ((value >> (bit + 1)) != 0) {
      bit++;
    }
    return bit;
  }

  private static int partition(int[] arr, int p, int r, int bit) {
    int i = p - 1;
    int j = r + 1;
    while (true) {
      do {
        i++;
      } while (i <= r && ((arr[i] >> bit) & 1) == 0);
      do {
        j--;
      } while (j >= p && ((arr[j] >> bit) & 1) == 1);
      if (i < j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
      } else {
        return j;
      }
    }
  }

  private static void binaryQuickSortRecursive(int[] arr, int p, int r, int bit) {
    if (p < r && bit >= 0) {
      int q = partition(arr, p, r, bit);
      binaryQuickSortRecursive(arr, p, q, bit - 1);
      binaryQuickSortRecursive(arr, q + 1, r, bit - 1);
    }
  }

  public static void sort(int[] arr) {
    int n = arr.length;
    int maxValue = arr[0];
    for (int i = 1; i < n; i++) {
      if (arr[i] > maxValue) {
        maxValue = arr[i];
      }
    }
    int bit = mostSignificantBit(maxValue);
    binaryQuickSortRecursive(arr, 0, n - 1, bit);
  }

  public static void main(String[] args) {
    int[] array = new int[] {0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56};
    sort(array);
    System.out.println(Arrays.toString(array));
  }
}

File: Modules/BuiltInAlgorithms/Sources/Distribution/BinaryQuickSortRecursive.swift
import AlgorithmKit
import SortEngineKit

/// Ported from ArrayV's `BinaryQuickSortRecursive` — a pure entry-point wrapper around
/// `BinaryQuickSortingTemplate.binaryQuickSortRecursive`'s call-stack driver. Same partition and
/// bit-depth-bounded recursion as `BinaryQuickSortIterative`, just recursing directly instead of
/// working off an explicit task queue — see `BinaryQuickSortingTemplate`'s own doc comment for the
/// shared algorithm.
public struct BinaryQuickSortRecursive: SortAlgorithm {
  public let id = AlgorithmID(rawValue: "binaryquicksortrecursive")
  public let metadata = AlgorithmMetadata(
    displayName: "Binary Quick Sort (Recursive)",
    category: .distribution,
    sizeRange: 16...256,
    growthModel: OperationGrowthModel(
      anchorSize: 2823, coefficients: [181627, 73.6464, 0.00170668],
      measuredSafeCeiling: nil),
    detectedGrowthModel: DetectedGrowthModel(
      family: .powerLog, coefficients: [6.97262, 1.01882], rSquared: 0.997448),
    implementationComplexity: 13,
    stable: false,
    timeComplexity: ComplexityBounds(
      best: "O(n log n)", average: "O(n log n)", worst: "O(n log n)"),
    spaceComplexity: "O(log n)",
    iconName: "square.stack.3d.up.badge.a"
  )

  public init() {}

  public func record(into engine: inout RecordingEngine) {
    let n = engine.count
    guard n > 1 else { return }
    let msb = BinaryQuickSortingTemplate.mostSignificantBit(engine.values)
    BinaryQuickSortingTemplate.binaryQuickSortRecursive(&engine, 0, n - 1, msb)
  }
}
